File: android/src/main/java/com/nabidreams/lib/audio/PlayerModule.kt
package com.nabidreams.lib.audio

import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.io.IOException

class PlayerModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    enum class EventType(val value: String) {
        STATE_CHANGE("playerStateChange")
    }

    override fun getName(): String {
        return "Player"
    }

    override fun getConstants(): MutableMap<String, Any> {
        return mapOf(
                "State" to Player.State.values().map { it.name to it.value }.toMap(),
                "EventType" to EventType.values().map { it.name to it.value }.toMap(),
                "MIN_LEVEL" to Player.MIN_POWER,
                "MAX_LEVEL" to Player.MAX_POWER
        ).toMutableMap()
    }

    @ReactMethod
    fun getLevel(promise: Promise) {
        promise.resolve(player.rmsPower)
    }

    @ReactMethod
    fun getState(promise: Promise) {
        promise.resolve(player.state.value)
    }

    @ReactMethod
    fun start(filePath: String, promise: Promise) {
        try {
            player.start(filePath)
            promise.resolve(null)
        } catch (e: IOException) {
            promise.reject(e)
        }
    }

    @ReactMethod
    fun stop(promise: Promise) {
        player.stop()
        promise.resolve(null)
    }

    private val player: Player = Player().apply {
        stateChangeListener = { state ->
            sendEvent(EventType.STATE_CHANGE.value, Arguments.makeNativeMap(mapOf("state" to state.value)))
        }
    }

    private fun sendEvent(eventName: String, data: Any? = null) {
        reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java).emit(eventName, data)
    }
}

File: ios/PlayerModule.swift
@objc(Player)
class PlayerModule: RCTEventEmitter {
  enum EventType: String {
    case stateChange = "playerStateChange"
  }
  
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
  
  override func constantsToExport() -> [AnyHashable: Any]? {
    return [
      "State": [
        "STARTED": Player.State.started.rawValue,
        "STOPPED": Player.State.stopped.rawValue,
      ],
      
      "EventType": [
        "STATE_CHANGE": EventType.stateChange.rawValue
      ],
      
      "MIN_LEVEL": Player.minPower,
      "MAX_LEVEL": Player.maxPower,
    ]
  }
  
  override func supportedEvents() -> [String]! {
    return [
      EventType.stateChange.rawValue
    ]
  }
  
  override init() {
    super.init()
    
    player.stateChangeListener = { state in
      self.sendEvent(withName: EventType.stateChange.rawValue, body: ["state": state.rawValue])
    }
  }
  
  @objc
  func getLevel(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    resolve(player.averagePower)
  }
  
  @objc
  func getState(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    resolve(player.state.rawValue)
  }
  
  @objc
  func start(_ filePath: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    do {
      try player.start(filePath)
      resolve(nil)
    } catch {
      reject("Error", error.localizedDescription, error)
    }
  }
  
  @objc
  func stop(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    player.stop()
    resolve(nil)
  }
  
  private let player: Player! = Player()
}
